File: src/UnionFind.rs
struct UnionFind {
    par: Vec<isize>,
    n: usize
}
 
impl UnionFind {
    fn init(length: usize) -> UnionFind {
        return UnionFind{par: vec![-1; length], n: length};
    }
    fn root(&mut self, x: usize) -> usize{
        if self.par[x] < 0 {
            return x;
        } else {
            self.par[x] = self.root(self.par[x] as usize) as isize;
            return self.par[x] as usize
        }
    }
    fn same(&mut self, x: usize, y: usize) -> bool {
        return self.root(x) == self.root(y);
    }
    fn merge(&mut self, x: usize, y: usize) -> bool{
        let mut rx = self.root(x);
        let mut ry = self.root(y);
        if rx == ry {return false;};
        if self.par[rx] > self.par[ry] {
            std::mem::swap(&mut rx, &mut ry);
        }
        self.par[rx] += self.par[ry];
        self.par[ry] = rx as isize;
        return true;
    }
    fn size(&mut self, x: usize) -> usize {
        let p = self.root(x);
        return -self.par[p] as usize;
    }
    fn status(&mut self) -> Vec<usize> {
        let mut sta = Vec::new();
        for i in 0..self.n {
            sta.push(self.root(i));
        }
        return sta;
    }
    fn islands(&mut self, begin: usize, end: usize) -> usize {
        let sta = self.status();
        let set = sta.iter().cloned().collect::<std::collections::HashSet<usize>>();
        return set.len();
    }
}

File: cpp/UnionFind.cpp
/*
Union-Find木
初期化（自動): init(頂点数)
頂点xの根を求める: root(頂点x)
頂点xと頂点yの木を結合: merge(頂点x, 頂点y)
頂点xと頂点yが同じ木に属するか: same(頂点x, 頂点y)
頂点xの属する木の頂点数: size(頂点x)
森に生えてる木の本数: forest_size(頂点x)

verified: https://judge.yosupo.jp/submission/56611
*/

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

struct UnionFind {
    int flen;
    vector<int> par;
    UnionFind(int n) {init(n);}
    void init(int n) {
        par.assign(n, -1);
        flen = n;
    }
    int root(int x) {
        if (par[x] < 0) return x;
        return par[x] = root(par[x]);
    }
    bool merge(int x, int y) {
        x = root(x);
        y = root(y);
        if (x == y) return false;
        if (par[x] > par[y]) swap(x, y);
        par[x] += par[y];
        par[y] = x;
        flen--;
        return true;
    }
    bool same(int x, int y) {
        return root(x) == root(y);
    }
    int size(int x) {
        return -par[root(x)];
    }
    int forest_size() {
        return flen;
    }
};

int main() {
    int n, q;
    cin >> n >> q;
    UnionFind uf(n);
    while (q--) {
        int p, a, b;
        cin >> p >> a >> b;
        if (p == 0) uf.merge(a, b);
        else cout << uf.same(a, b) << "\n";
    }
    return 0;
}
